src/librustc_resolve/lib.rs

```diff
@@ -2077,40 +2086,42 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             this.visit_generics(generics);
```

Comment:
would it possibly make sense to first do the `this.with_current_self_type(self_type, |this| { ... })` and only do the `this.visit_generics(generics);` within the `{ ... }`?  It seems like it might fix #24944